src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1755,206 +1906,299 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
 
         public static int Min(this IEnumerable<int> source)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            Error.CheckSourceNotNull(source);
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
```

Comment:
> > I haven't added short-circuits where there isn't a test in the original

> We have very low coverage of LINQ in our tests right now. Please don't infer that a lack of a test for a particular behavior means that behavior can be changed (though it'd be nice if we lived in such a world :smile:).

Sorry. What I meant wasn't *test* as in *test-case* but *test* as in the running code tests for a condition, and branches accordingly. We could for example add `if (value == int.MinValue) return value;` into `Min<int>()` and have a short-circuit that benefits sequences containing `int.MinValue` at the cost of adding extra work to those that don't and is probably not worth it. `Min<double>()` on the other hand already has to test for `double.IsNaN(value)` so making that a short-circuit condition doesn't add a test to the loop.